### ISSUE TITLE
Is trainable actually set to False in load_initial_weights?

### ISSUE THREAD

**Issue by francescomilano172** [kratzert/finetune_alexnet_with_tensorflow#70]
As already answered in #55, the variables for which one wants to load the values from the ImageNet weights (i.e., those not in `SKIP_LAYERS`) are set to `trainable=False` in `load_initial_weights`. This is done by setting `reuse=True` when using `tf.get_variable` on these variables, which were previously created either explicitly with `trainable=True` (e.g. for `fc`) or with no specified value for `trainable` (as in `conv`), which defaults to `trainable=True`, as can be seen from Tensorflow code (https://github.com/tensorflow/tensorflow/blob/a6d8ffae097d0132989ae4688d224121ec6d8f35/tensorflow/python/ops/variable_scope.py#L495 and https://github.com/tensorflow/tensorflow/blob/a6d8ffae097d0132989ae4688d224121ec6d8f35/tensorflow/python/ops/variable_scope.py#L2400).

Does setting `trainable=False` therefore really have an effect? When I display the list of `tf.trainable_variables()` I can see also those that are not in `SKIP_LAYERS`. Indeed, I thought that it is not possible to change the `trainable` property of a Variable after creation, as also shown in https://stackoverflow.com/a/37327561.

I know that the variables to be actually trained can still be specified when running the model, as done with `train_layers` in `finetune.py`, but I just wanted to make sure that I am not missing something.

**Comment by kratzert** [comment]
Tbh I'm not sure how TF handles this atm. When I designed this code I am pretty sure that this works to only train the specific layers mentioned in the train layers list. However, I left TF quite a long time ago and only use PyTorch now. And since TF changes the API frequently it might well be that now all layers are trained.

**Comment by francescomilano172** [comment]
Thank you for your answer. It is still the case that only variables from layers in `train_layers` are trained, thanks to the definition of `var_list` in https://github.com/kratzert/finetune_alexnet_with_tensorflow/blob/669f32226a2f1d0669a81824df99758f0edad557/finetune.py#L92
Just for the sake of precision, `tf.trainable_variables()` in the same line will also contain the variables that have been 'set' to `trainable=False` (with no effect), such as
https://github.com/kratzert/finetune_alexnet_with_tensorflow/blob/669f32226a2f1d0669a81824df99758f0edad557/alexnet.py#L116
But I guess what matters is that only layers in `train_layers` are trained, so as long as gradients are only generated and applied for the `var_list` above this shouldn't be too big a problem.
